internal/proxy/task_policies.go

```diff
@@ -14,35 +14,36 @@ import (
 
 // type pickShardPolicy func(ctx context.Context, mgr *shardClientMgr, query func(UniqueID, types.QueryNode) error, leaders []nodeInfo) error
 
-type pickShardPolicy func(context.Context, *shardClientMgr, func(context.Context, UniqueID, types.QueryNode, []string) error, map[string][]nodeInfo) error
+type pickShardPolicy func(context.Context, *shardClientMgr, func(context.Context, UniqueID, types.QueryNode, []string) error, ShardLeaders) error
 
 var (
 	errBegin               = errors.New("begin error")
 	errInvalidShardLeaders = errors.New("Invalid shard leader")
 )
 
-func updateShardsWithRoundRobin(shardsLeaders map[string][]nodeInfo) {
-	for channelID, leaders := range shardsLeaders {
-		if len(leaders) <= 1 {
+func updateShardsWithRoundRobin(shardsLeaders ShardLeaders) {
+	for items := range shardsLeaders.IterBuffered() {
+		if len(items.Val) <= 1 {
 			continue
 		}
-
-		shardsLeaders[channelID] = append(leaders[1:], leaders[0])
+		items.Val = append(items.Val[1:], items.Val[0])
+		shardsLeaders.Set(items.Key, items.Val)
```

Comment:
If we need COW here, we shall clone the values of items and set it back instead of writing it directly.

Reply:
solved